src/main/kotlin/slack/cli/exec/ProcessedExecCli.kt

```diff
@@ -0,0 +1,151 @@
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package slack.cli.exec
+
+import com.github.ajalt.clikt.core.CliktCommand
+import com.github.ajalt.clikt.parameters.arguments.argument
+import com.github.ajalt.clikt.parameters.arguments.multiple
+import com.github.ajalt.clikt.parameters.options.flag
+import com.github.ajalt.clikt.parameters.options.option
+import com.github.ajalt.clikt.parameters.types.path
+import com.squareup.moshi.adapter
+import eu.jrie.jetbrains.kotlinshell.shell.shell
+import java.nio.file.Path
+import kotlin.io.path.ExperimentalPathApi
+import kotlin.io.path.createDirectories
+import kotlin.io.path.createTempFile
+import kotlin.io.path.deleteRecursively
+import kotlin.system.exitProcess
+import okio.buffer
+import okio.source
+import slack.cli.projectDirOption
+
+/**
+ * Executes a command with Bugsnag tracing and retries as needed.
+ *
+ * Example:
+ * ```
+ * $ ./<binary> --bugsnag-key=1234 --verbose --configurationFile config.json ./gradlew build
+ * ```
+ */
+public class ProcessedExecCli :
+  CliktCommand("Executes a command with Bugsnag tracing and retries as needed.") {
+
+  private val projectDir by projectDirOption()
+  private val verbose by option("--verbose", "-v").flag()
+  private val bugsnagKey by option("--bugsnag-key", envvar = "PE_BUGSNAG_KEY")
+  private val configurationFile by
+    option("--config", envvar = "PE_CONFIGURATION_FILE")
+      .path(mustExist = true, canBeFile = true, canBeDir = false)
+
+  private val args by argument().multiple()
+
+  @OptIn(ExperimentalStdlibApi::class, ExperimentalPathApi::class)
+  override fun run() {
+    val moshi = ProcessingUtil.newMoshi()
+    val config =
+      configurationFile?.let {
+        echo("Parsing config file '$it'")
+        it.source().buffer().use { source -> moshi.adapter<ProcessedExecConfig>().fromJson(source) }
+      }
+        ?: ProcessedExecConfig()
+    // The command to be executed
+    val cmd = args.joinToString(" ")
+
+    // Temporary file for command output
+    val tmpDir = projectDir.resolve("tmp/processed_exec")
+    tmpDir.createDirectories()
+
+    // Initial command execution
+    val (exitStatus, logFile) = executeCommand(cmd, tmpDir)
+    while (exitStatus != 0) {
+      echo("Command failed with exit code $exitStatus. Running processor script...")
+
+      echo("Processing CI failure")
+      val resultProcessor = ResultProcessor(verbose, bugsnagKey, config, ::echo)
+
+      when (val retrySignal = resultProcessor.process(logFile, false)) {
+        is RetrySignal.Ack,
+        RetrySignal.Unknown -> {
+          echo("Processor exited with 0, exiting with original exit code...")
```

Comment:
This output is wrong, if it exited with 0 we wouldn't be here. We also exit this process with 0 _always_ (line 111), do we want to reflect the actual result of the command or do we always want to succeed?

Reply:
Good catch [e8c3d17](https://github.com/slackhq/kotlin-cli-util/pull/20/commits/e8c3d175de04373b0e751688325e41eddd1913f2)